Feature/12 last events

e-floran opened the pull request:
Ajout de l'onglet "Last events" qui repertorie les évènements passés depuis la dernière visite de l'utilisateur.

e-floran committed:
first commit

e-floran committed:
working on date

e-floran committed:
app is working but not the filter

e-floran committed:
solved filter problem

e-floran committed:
merged dev

e-floran committed:
added Redux

clovfr commented on src/App.jsx:
```diff
@@ -1,12 +1,28 @@
 function App() {
+  const [lastVisit, setLastVisit] = useState('');
+  const [storageCount, setStorageCount] = useState(0);
+  function handleCount() {
+    setTimeout(function () {
```
eslint râle : utilises des fonctions flèches

e-floran replied:
Ah mince, il ne disait rien chez moi...

clovfr replied:
Après application du patch (je ne sais pas si c'est lié) le paramètre pour choisir le linter/prettier a sauté dans OSS/VCS. Vérifie si c'est toujours bon chez toi.

clovfr commented on src/App.jsx:
```diff
@@ -1,12 +1,28 @@
+  useEffect(() => {
+    if (window.localStorage.getItem('lastVisit')) {
+      setLastVisit(window.localStorage.getItem('lastVisit').slice(0, 10));
+      console.log(lastVisit);
```
Pense à cleaner avant de PR ;)

clovfr commented on src/App.jsx:
```diff
@@ -1,12 +1,28 @@
+      console.log(lastVisit);
+    }
+    window.localStorage.setItem('lastVisit', new Date().toISOString());
+    console.log(window.localStorage.getItem('lastVisit'));
```
clean

clovfr commented on src/App.jsx:
```diff
@@ -1,12 +1,28 @@
 
 function App() {
+  const [lastVisit, setLastVisit] = useState('');
+  const [storageCount, setStorageCount] = useState(0);
+  function handleCount() {
+    setTimeout(function () {
+      setStorageCount(storageCount + 1);
+    }, 15000);
+  }
+  handleCount();
+  useEffect(() => {
+    if (window.localStorage.getItem('lastVisit')) {
+      setLastVisit(window.localStorage.getItem('lastVisit').slice(0, 10));
+      console.log(lastVisit);
+    }
+    window.localStorage.setItem('lastVisit', new Date().toISOString());
+    console.log(window.localStorage.getItem('lastVisit'));
+  }, [storageCount]);
```
Le timeout fait rafraîchir `lastVisit` dans localStorage toutes les 15secondes, du coup la requête API sera toujours équivalente à `{start: now(), end: now()}`.

clovfr reviewed (changes requested):
L'implantation de redux dans le composant est bonne, mais la méthode pour détecter une précédente visite n'est pas bonne.

clovfr commented on src/App.jsx:
```diff
@@ -1,12 +1,28 @@
+      setLastVisit(window.localStorage.getItem('lastVisit').slice(0, 10));
+      console.log(lastVisit);
+    }
+    window.localStorage.setItem('lastVisit', new Date().toISOString());
```
il manque un `else` en début de ligne (s'il n'y a pas de valeur, alors la set). Mais il faudrait passer cette nouvelle valeur dans `setLastVisit` aussi.

e-floran replied:
J'ai rajouté le else mais pas le passage à setLastVisit, puisque le but est de savoir si l'utilisateur est déjà venu, justement. Dans le cas contraire, le state doit être vide !

e-floran committed:
fixed PR needed adjustments

clovfr commented on src/App.jsx:
```diff
@@ -5,21 +5,13 @@ import Map from './components/Map/Map';
-    window.localStorage.setItem('lastVisit', new Date().toISOString());
-    console.log(window.localStorage.getItem('lastVisit'));
-  }, [storageCount]);
+  }, []);
```
J'ai maintenant une erreur quand je charge la page avec un localStorage vierge.
Vérifie la présence de l'entré dans localStorage en premier et si elle n'existe pas, créé-la. Ensuite charge-la tout simplement.

*in a nutshell* : 
```js
if (!window.localStorage.getItem('lastVisit')) {
  window.localStorage.setItem('lastVisit', new Date().toISOString());
}
setLastVisit(window.localStorage.getItem('lastVisit').slice(0, 10));
```

clovfr replied:
Par contre, dans le render
```JS
{lastVisit === '' && (
  <p className="error">This is your first visit, please come again.</p>
)}
```
ne s'exécutera jamais

e-floran committed:
working on localStorage fix

Commit:
func App to class App

lastVisit must be set earlier than componentDidMount (via useEffect) or the Map will crash (bad bounds ? overloaded map ? idk, using class component just did fix it for now).

e-floran commented:
C'est tout bon ?

ghost commented:
LGTM
